src/Abstract/Value.hs

```diff
@@ -78,23 +78,25 @@ instance Ord l => AbstractValue l (Value l) where
 
   literal = I
 
-instance (MonadStore Monovariant Type m, MonadEnv Monovariant Type m, MonadFail m, Semigroup (Cell Monovariant Type), Alternative m) => MonadValue Monovariant Type t m where
-  rec ev name ty e0 =  do
+instance (MonadStore Monovariant Type m, MonadEnv Monovariant Type m, MonadFail m, Semigroup (Cell Monovariant Type), MonadFresh m, Alternative m) => MonadValue Monovariant Type t m where
+  rec ev name e0 =  do
     a <- alloc name
-    assign a ty
+    tvar <- fresh
+    assign a (TVar tvar)
     v <- localEnv (envInsert name (a :: Address Monovariant Type)) (ev e0)
     return v
 
-  lambda ev name inTy body = do
+  lambda ev name body = do
     a <- alloc name
-    assign a inTy
+    tvar <- fresh
+    assign a (TVar tvar)
     outTy <- localEnv (envInsert name (a :: Address Monovariant Type)) (ev body)
-    return (inTy :-> outTy)
+    return (TVar tvar :-> outTy)
 
-  app _ (inTy :-> outTy) argTy = do
-    unless (inTy == argTy) (fail ("expected " ++ show inTy ++ " but got " ++ show argTy))
+  app _ opTy inTy = do
+    tvar <- fresh
+    _ :-> outTy <- opTy `unify` (inTy :-> TVar tvar)
```

Comment:
Rather than explicitly error-checking applications, we unify the operator’s type against a function type. We’ll still get error messages if it doesn’t unify, and this way we’ve got fewer code paths generating them.